Docs: Add indexing guide and index advisor extension (#16238)
supabase/supabase: apps/docs/pages/guides/database/query-optimization.mdx

import Layout from '~/layouts/DefaultGuideLayout'

export const meta = {
  title: 'Query Optimization',
  description: 'Choosing indexes',
  footerHelpType: 'postgres',
}

When working with PostgreSQL, or any relational database, indexing is key to improving query performance.
Aligning indexes with common query patterns can speed up data retrieval by order of magnitude.

This guide is intended to:

- help identify parts of a query that have the potential to be improved by indexes

> - introduce the most frequently applicable indexing tool

Author: resolved in f8aa35dd4

```sql
create index idx_orders_created_at ON customers using brin(created_at);
```

### Partial Indexes

For queries that frequently target a subset of data, a partial index could be faster and smaller than indexing the entire column. A partial index contains a `where` clause to filter the values included in the index. Note that a query's `where` clause must match the index for it to be used.

```sql
create index idx_orders_status on orders (status)
where status = 'shipped';
```

### Composite Indexes

If filtering or joining on multiple columns, a composite index can help postgres referring to multiple indexes to identify the relevant rows.

```sql
create index idx_customers_sign_up_date_priority on customers (sign_up_date, priority);
```

### Over-Indexing

Avoid the urge to index columns you operate on infrequently. While indexes can speed up reads, they also slow down writes, so its important to balance those factors when making indexing decisions.

### Statistics

> Postgres maintains a set of statistics about the contents of your tables. Those statistics are used by the query planner to decide e.g. when its is more efficient to use an index vs scanning the entire table. If the collected statistics drift too far from reality, the query planner may make poor decisions. To avoid the risk, you can periodically `analyze` tables.

Reviewer: ```suggestion
Postgres maintains a set of statistics about the contents of your tables. Those statistics are used by the query planner to decide when it's is more efficient to use an index vs scanning the entire table. If the collected statistics drift too far from reality, the query planner may make poor decisions. To avoid this risk, you can periodically `analyze` tables.
```
